### ISSUE TITLE
How can I test yolov5 weight that I custom train

### ISSUE THREAD

**Issue by bonlycpe** [ultralytics/yolov5#8070]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

How can I test yolov5 weight that I custom train my self

**Comment by bonlycpe** [comment]
I have found this one
![image](https://user-images.githubusercontent.com/106681968/171425525-addf3343-de44-45e6-9ca9-323c28f057f2.png)

**Comment by glenn-jocher** [comment]
@conandoor1 
```
python val.py --data DATA.yaml --weights MODEL.pt
```

**Comment by bonlycpe** [comment]
thank ^^ , And another question I have see that in train.py have use val.py too that mean every train it will automatically test in the end right , I'm confused about optional in this tutorials  that have test too.

![image](https://user-images.githubusercontent.com/106681968/171576974-98834ea2-1465-410c-99c6-181553c9e0f8.png)

**Comment by glenn-jocher** [comment]
@conandoor1 yes train.py runs validation every epoch and after training completes on best.pt.

**Comment by bonlycpe** [comment]
@glenn-jocher Thank ^^

**Comment by bonlycpe** [comment]
@glenn-jocher how can I get result like in tutorial I away got like this 
![image](https://user-images.githubusercontent.com/106681968/171633983-a3bc0cdd-b992-4e5c-a7ef-245b38712c36.png)
![image](https://user-images.githubusercontent.com/106681968/171633998-1e936e93-83d2-4758-a7fa-edbb6c87b62f.png)

**Comment by glenn-jocher** [comment]
@conandoor1 the second pic you have there is showing pycocotools results. These are only applied when evaluating the COCO dataset.

**Comment by bonlycpe** [comment]
@glenn-jocher Thank ^^

**Comment by pabsan-0** [comment]
@glenn-jocher Hello! I get wildly different AP results when comparing `val.py` and testing myself with pycocotools:
- Approach 1: `python3 val.py ...`
```
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 32/32 [00:49<00:00,  1.55s/it]                  
                 all       1024        678      0.546      0.494      0.518      0.221
Speed: 0.4ms pre-process, 2.7ms inference, 0.5ms NMS per image at shape (32, 3, 1280, 1280)
                                                                        *          *
                                                                        |          |
```
- Approach 2: `python3 val.py --save-json ...`; then adjust the json and use pycocotools to get average-recall tables.
```
Accumulating evaluation results...
DONE (t=0.03s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.127 *---------
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.323 *---------
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.081
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.039
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.083
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.191
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.177
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.186
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.186
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.124
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.174
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.218
```

Could you please shed some light on why is this happening and whether it is normal? 

I've had this scame scenario happen too with YoloR. Back in the day I got the same-ish custom (6 class) AP results with pycocotools and darknet... pycocotools repo hasnt changed ever since so it is my understanding pycocotools metrics should be fine for custom data too.

**Comment by glenn-jocher** [comment]
@pabsan-0 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by pabsan-0** [comment]
@glenn-jocher Hello! Thanks for your reply. After reading [your previous answer in this thread](https://github.com/ultralytics/yolov5/issues/8070#issuecomment-1144990316) I though you discouraged COCO metrics for non-COCO datasets and I assumed either yolov5 somehow just would not play nice with external COCO validation or that this diff was my data's fault.

Then I saw this other issue in which you state [--conf-thresh should be 0.001](https://github.com/ultralytics/yolov5/issues/1952#issuecomment-761899820). I made this correction and both validation approaches seem to yield the same results now (with the usual 1% diff). Will raise my own issue if something alike comes up again. 

Cheers!
